### PR TITLE
fix unexpected append mutations about pkg/controller packages

### PR DESCRIPTION
**Pull request by kadisi**
Co-authored-by: Dr. Stefan Schimanski <stefan.schimanski@gmail.com>
Signed-off-by: kadisi <iamkadisi@163.com>

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

/kind bug

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

refs #67815

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by kadisi** [comment]
cc @sttts @RainbowMango

File: pkg/controller/deployment/recreate.go [hunk 1]
```diff
@@ -31,7 +31,7 @@ func (dc *DeploymentController) rolloutRecreate(d *apps.Deployment, rsList []*ap
 	if err != nil {
 		return err
 	}
-	allRSs := append(oldRSs, newRS)
+	allRSs := append(append([]*apps.ReplicaSet{nil}, oldRSs...), newRS)
```
Comment (RainbowMango) [inline pkg/controller/deployment/recreate.go:34]: `[]*apps.ReplicaSet{nil}` will add a nil element that we don't expected to the slice.
Same as below.

**Comment by RainbowMango** [comment]
/ok-to-test
/priority backlog
Thanks @kadisi

File: pkg/controller/deployment/recreate.go (hunk 1 above)
Comment (kadisi) [reply pkg/controller/deployment/recreate.go]: done

File: pkg/controller/deployment/recreate.go [hunk 2]
```diff
@@ -18,7 +18,7 @@ package deployment
 
 import (
 	apps "k8s.io/api/apps/v1"
-	"k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
```
Comment (tnozicka) [inline pkg/controller/deployment/recreate.go:21]: nit: corev1

File: pkg/controller/deployment/recreate.go [hunk 3]
```diff
@@ -31,7 +31,7 @@ func (dc *DeploymentController) rolloutRecreate(d *apps.Deployment, rsList []*ap
 	if err != nil {
 		return err
 	}
-	allRSs := append(oldRSs, newRS)
+	allRSs := append(append([]*apps.ReplicaSet{}, oldRSs...), newRS)
```
Comment (tnozicka) [inline pkg/controller/deployment/recreate.go:34]: Would it be better to pre-allocate the array with `make` capacity=len(oldRs), then declaring it inline, and then 2 appends for oldRSs and newRS?

File: pkg/controller/deployment/recreate.go (hunk 2 above)
Comment (kadisi) [reply pkg/controller/deployment/recreate.go]: done

File: pkg/controller/deployment/recreate.go (hunk 3 above)
Comment (kadisi) [reply pkg/controller/deployment/recreate.go]: done

**Comment by kadisi** [comment]
@tnozicka PTAL

File: pkg/controller/deployment/recreate.go [hunk 4]
```diff
@@ -18,20 +18,22 @@ package deployment
...
-func (dc *DeploymentController) rolloutRecreate(d *apps.Deployment, rsList []*apps.ReplicaSet, podMap map[types.UID][]*v1.Pod) error {
+func (dc *DeploymentController) rolloutRecreate(d *apps.Deployment, rsList []*apps.ReplicaSet, podMap map[types.UID][]*corev1.Pod) error {
 	// Don't create a new RS if not already existed, so that we avoid scaling up before scaling down.
 	newRS, oldRSs, err := dc.getAllReplicaSetsAndSyncRevision(d, rsList, false)
 	if err != nil {
 		return err
 	}
-	allRSs := append(oldRSs, newRS)
+	tmpRs := make([]*apps.ReplicaSet, len(oldRSs))
+	copy(tmpRs, oldRSs)
```
Comment (tnozicka) [inline pkg/controller/deployment/recreate.go:35]: I was thinking more:
```
allRSs := make([]*apps.ReplicaSet, 0, len(oldRSs)+1)
allRSs = append(allRSs, oldRSs)
allRSs = append(allRSs, newRS)
```

Comment (kadisi) [reply pkg/controller/deployment/recreate.go:35]:
```
allRSs := make([]*apps.ReplicaSet, 0, len(oldRSs)+len(newRS))
copy(allRSs, oldRSs)
allRSs = append(allRSs, newRS)
```
is it ok ?

**Pull request closed without merging**